src/commands/update.rs

```diff
@@ -1,6 +1,146 @@
+                    // NOTE: A bit of an edge case. If the channel is present in
+                    // the local manifest but not in upstream, then it probably
+                    // either:
+                    // - is a developer toolchain.
+                    // - the upstream channel got removed from upstream (possibly for being too
+                    //   old/deprecated/got rolled back)
+                    continue;
+                };
+                channels_to_update.push((local_channel.clone(), upstream_channel.clone()));
+            }
+
+            for (local_channel, upstream_channel) in channels_to_update {
+                update_channel(config, &local_channel, &upstream_channel, local_manifest)?;
+            }
+        },
+        Some(UserChannel::Nightly) => todo!(),
+        Some(UserChannel::Other(_)) => todo!(),
+    }
+    Ok(())
+}
+
+fn update_channel(
+    config: &Config,
+    local_channel: &Channel,
+    upstream_channel: &Channel,
+    local_manifest: &mut Manifest,
+) -> anyhow::Result<()> {
+    let installed_toolchains_dir = config.midenup_home.join("toolchains");
+    let toolchain_dir = installed_toolchains_dir.join(format!("{}", &local_channel.name));
+
+    let updates = local_channel.components_to_update(upstream_channel);
+
+    let libs = ["std", "base"];
+    let (libraries, executables): (Vec<_>, Vec<_>) =
+        updates.iter().partition(|c| libs.contains(&(c.name.as_ref())));
+
+    for lib in libraries {
+        let lib_path = toolchain_dir.join("lib").join(lib.name.as_ref()).with_extension("masp");
+        std::fs::remove_file(&lib_path)
+            .context(format!("Couldn't delete {}", &lib_path.display()))?;
+    }
+
+    let toolchain_dir = config
+        .midenup_home
+        .join("toolchains")
+        .join(format!("{}", &upstream_channel.name));
+
+    for exe in executables {
+        match &exe.version {
+            Authority::Cargo { package, .. } => {
+                let mut remove_exe = std::process::Command::new("cargo")
+                    .arg("uninstall")
+                    .arg(package.as_deref().unwrap_or(exe.name.as_ref()))
+                    .arg("--root")
+                    .arg(&toolchain_dir)
+                    .stderr(std::process::Stdio::inherit())
+                    .stdout(std::process::Stdio::inherit())
+                    .spawn()
+                    .with_context(|| {
+                        format!(
+                            "failed to uninstall {} via cargo",
+                            package.as_deref().unwrap_or(exe.name.as_ref())
+                        )
+                    })?;
+
+                remove_exe
+                    .wait()
+                    .context("failed to uninstall component '{{ component.name }}'")?;
+            },
+            _ => todo!(),
+        }
+    }
+
+    // NOTE: After deleting the files we need to remove the "all is installed
```

Comment:
I would remove this file first actually - that way if something goes wrong, a subsequent `install`/`update` will reinstall everything. If we remove it last, we might remove various components, hit an error part way through, and then bail before removing the install marker.